Flaky test (deadlock): pkg/logging TestConcurrentWriteAndReloadIsRaceFree hangs 10m on windows-latest

## Summary

`TestSink/TestConcurrentWriteAndReloadIsRaceFree` in `pkg/logging` intermittently **deadlocks on `windows-latest`**, hanging the whole package for the default 10‑minute `go test` timeout and panicking the run. Unlike the watcher flake, this is **not** a tight-timeout issue — it is a genuine Windows-specific hang in the file-close path, surfaced by the test's concurrent-write + reload stress loop.

The same close path runs on the **production SIGHUP reload** (`Sink.Reload → applyDestination → oldFile.Close()`), so this is a latent Windows risk for a heavily-logging server during a `log_file` rotation, not purely a test artifact.

## Evidence

Two independent failures on `main`:

- https://github.com/containers/kubernetes-mcp-server/actions/runs/27294128498/job/80622565341
- https://github.com/containers/kubernetes-mcp-server/actions/runs/27017672308/job/79737182094

```
panic: test timed out after 10m0s
  running tests:
    TestSink/TestConcurrentWriteAndReloadIsRaceFree (10m0s)
FAIL  github.com/containers/kubernetes-mcp-server/pkg/logging  600.115s
```

Goroutine dump (abridged, from run 27017672308):

```
goroutine 25 [semacquire, 9 minutes]:          // the reload loop
  internal/poll.(*FD).Close(0xc00026a000)
  os.(*File).Close(...)
  logging.(*Sink).applyDestination  sink.go:243
  logging.(*Sink).Reload            sink.go:168
  ...TestConcurrentWriteAndReloadIsRaceFree  sink_test.go:303

goroutine 27 [runnable]:                        // a klog writer, HOLDS klog's global mutex
  syscall.WriteFile(0x240, ...)
  internal/poll.(*FD).Write(0xc00026a0c0, ...)
  os.(*File).Write(...)
  logging.(*Sink).Write             sink.go:141
  k8s.io/klog/v2.(*loggingT).output klog.go:913   // past the l.mu.Lock() at :887

goroutine 26 [sync.Mutex.Lock]:                 // another klog writer, blocked on klog's mutex
  k8s.io/klog/v2.(*loggingT).output klog.go:887

goroutine 28/29 [runnable]:                     // remaining hot-loop writers
  sink_test.go:290 / :291
```

## Root cause

`TestConcurrentWriteAndReloadIsRaceFree` (`sink_test.go:267-307`) runs 4 goroutines logging through klog in a tight hot loop while the main goroutine performs 1000 `Reload`s, each of which opens a new file and **closes the previously-open one** on the hot path.

Three facts combine into a Windows-only deadlock:

1. **klog serializes every log call through one global `loggingT.mu`, held across the backend write.** The dump shows goroutine 27 inside `Sink.Write` → blocking `syscall.WriteFile` while holding `loggingT.mu` (it is at `klog.go:913`, past the lock at `:887`); goroutine 26 is parked acquiring that same mutex. So at most one writer is ever inside `Sink.Write`.
2. **The Sink closes a just-rotated file inline while writes may still be routing through the atomic writer pointer.** This overlap is explicitly accepted in the code for POSIX (`sink.go:238-240`: "In-flight writes that already loaded the old writer may still hit the previous fd briefly after Close"). On POSIX a write to a closing fd returns fast (`EBADF`); on Windows the `internal/poll` close path interacts with the runtime poller / IOCP differently.
3. **The reloader is wedged in `os.File.Close()` → `internal/poll.(*FD).Close` → `runtime_Semacquire` for the full 10 minutes** (`sink.go:243`).

Net effect: the file-close path and the always-on concurrent klog writers deadlock on Windows. (Note: at the dump instant the in-flight `WriteFile` is on a *different* fd than the one `Close` is stuck on, so this is a close-vs-concurrent-writers wedge in the Windows poller rather than a single fd ownership chain — the stress loop makes the rare wedge reliably reachable.)

`make test` runs `go test -race -count=1 -v ./...` with **no `-timeout`**, so the default 10‑minute per-package timeout is what finally panics the run.

## Proposed fix

**Preferred (fixes the latent product bug):** guard `Write` against the swap+close so no write/poller op can overlap the descriptor being closed — e.g. a `sync.RWMutex` where `Write` takes `RLock` around `h.Write(p)` and `applyDestination` takes the exclusive `Lock` around `s.file.Swap(...)` + `oldFile.Close()`. This removes the Windows hang **and** closes the deliberately-accepted use-after-close race at `sink.go:238-243`.

Tradeoff: this adds a lock to the write path, against the package's current lock-free atomic-pointer design (see package doc). In practice an `RLock` is cheap and uncontended; only `Reload` (rare — SIGHUP) takes the exclusive lock, so the hot logging path is effectively unaffected.

**Lighter mitigations (test resilience only, do not fix the product path):**
- Add an explicit `-timeout` to `make test` so a hang fails fast instead of burning 10 minutes.
- Reduce the test's iteration count / gate the stress loop — but this leaves the Windows close behavior unaddressed.

## Acceptance criteria

- `TestSink/TestConcurrentWriteAndReloadIsRaceFree` no longer hangs on `windows-latest` (passes reliably under `-race`).
- `Sink.Reload` cannot close a file while a concurrent `Sink.Write` is in flight (use-after-close race at `sink.go:238-243` eliminated).

---
_Triage/analysis assisted by Claude Code._

--- a/pkg/logging/sink.go
+++ b/pkg/logging/sink.go
@@ -6,7 +6,7 @@
 // that "modifying the logger is not thread-safe and should be done while no
 // other goroutines invoke log calls, usually during program initialization."
 // We honor that by configuring klog exactly once in New and routing every
-// subsequent reload through an atomic writer pointer that the textlogger
+// subsequent reload through a mutex-guarded writer that the textlogger
 // writes to.
 package logging
 
@@ -18,7 +18,6 @@ import (
 	"os"
 	"strconv"
 	"sync"
-	"sync/atomic"
 
 	"github.com/go-logr/logr"
 	"k8s.io/klog/v2"
@@ -35,14 +34,9 @@ const StderrSentinel = "stderr"
 // own -v flag is the real verbosity gate at runtime. See package docs.
 const maxTextloggerVerbosity = 9
 
-// writerHolder lets us put an io.Writer (an interface) inside an
-// atomic.Pointer without the awkward double indirection of
-// atomic.Pointer[io.Writer].
-type writerHolder struct{ io.Writer }
-
 // Sink is the runtime-reloadable log destination. It implements io.Writer
-// so the textlogger can write through it; reloads atomically swap the
-// underlying writer and close the previous file (if any).
+// so the textlogger can write through it; reloads swap the underlying
+// writer and close the previous file (if any).
 //
 // Construct with New. Reload on SIGHUP. Close on shutdown.
 type Sink struct {
@@ -68,14 +62,14 @@ type Sink struct {
 	// stay concurrent); reload and Close take the exclusive lock.
 	mu sync.RWMutex
 
-	// writer is what every Write call routes through. Updated by reload via
-	// Store; never mutated in place.
-	writer atomic.Pointer[writerHolder]
+	// writer is what every Write call routes through, guarded by mu. Replaced
+	// (never mutated in place) by reload and Close under the exclusive lock.
+	writer io.Writer
 
 	// file is the currently-open log file (nil when the destination is
-	// httpOut, errOut, or io.Discard). Owned by the Sink — close happens on
-	// reload (when replaced) and Close.
-	file atomic.Pointer[os.File]
+	// httpOut, errOut, or io.Discard), guarded by mu. Owned by the Sink —
+	// close happens on reload (when replaced) and Close.
+	file *os.File
 
 	// lastLogFile is the last cfg.LogFile value applyDestination acted on,
 	// used to short-circuit reloads that wouldn't change the destination
@@ -145,11 +139,10 @@ func New(cfg *config.StaticConfig, httpOut, errOut io.Writer) (*Sink, error) {
 func (s *Sink) Write(p []byte) (int, error) {
 	s.mu.RLock()
 	defer s.mu.RUnlock()
-	h := s.writer.Load()
-	if h == nil {
+	if s.writer == nil {
 		return io.Discard.Write(p)
 	}
-	return h.Write(p)
+	return s.writer.Write(p)
 }
 
 // SDKLogger is the slog.Logger to hand to the MCP SDK's ServerOptions. It
@@ -159,14 +152,14 @@ func (s *Sink) SDKLogger() *slog.Logger {
 }
 
 // Reload re-applies cfg to the running sink: it opens the new log file (if
-// any), atomically swaps the writer, closes the previous file, and updates
-// klog's verbosity.
+// any), swaps the writer, closes the previous file, and updates klog's
+// verbosity.
 //
 // Reload is intended to be called by a single goroutine (the SIGHUP
-// handler). Concurrent Reload calls would not corrupt klog state — each
-// step is individually atomic — but they could leave the sink pointing at
-// an already-closed file via a Store/Swap interleaving. The package
-// assumes a single reloader.
+// handler). mu makes each swap+close safe against concurrent Writes, but
+// two concurrent Reloads could still interleave their open-then-swap and
+// leave the sink pointing at an already-closed file. The package assumes
+// a single reloader.
 //
 // The returned error reflects only the destination swap (open-file
 // failure). On error, the previous destination is preserved unchanged.
@@ -197,8 +190,9 @@ func (s *Sink) Close() error {
 	// Exclusive lock so the swap+close can't race an in-flight Write (see mu).
 	s.mu.Lock()
 	defer s.mu.Unlock()
-	s.writer.Store(&writerHolder{s.errOut})
-	old := s.file.Swap(nil)
+	s.writer = s.errOut
+	old := s.file
+	s.file = nil
 	if old == nil {
 		return nil
 	}
@@ -206,9 +200,8 @@ func (s *Sink) Close() error {
 }
 
 // applyDestination resolves cfg.LogFile to a writer (and optionally an open
-// file), atomically swaps it in, and closes the previously-held file. It is
-// the only path that opens or closes log files — Reload and New both go
-// through it.
+// file), swaps it in, and closes the previously-held file. It is the only
+// path that opens or closes log files — Reload and New both go through it.
 //
 // A reload that wouldn't change the destination is skipped, with one
 // exception: a real file path always reopens, even when the path is
@@ -252,8 +245,9 @@ func (s *Sink) applyDestination(cfg *config.StaticConfig) error {
 	// above, outside the lock, so writers only ever block for the swap+close,
 	// not for open latency.
 	s.mu.Lock()
-	s.writer.Store(&writerHolder{newWriter})
-	oldFile := s.file.Swap(newFile)
+	s.writer = newWriter
+	oldFile := s.file
+	s.file = newFile
 	var closeErr error
 	if oldFile != nil {
 		closeErr = oldFile.Close()

--- a/pkg/logging/sink.go
+++ b/pkg/logging/sink.go
@@ -17,6 +17,7 @@ import (
 	"log/slog"
 	"os"
 	"strconv"
+	"sync"
 	"sync/atomic"
 
 	"github.com/go-logr/logr"
@@ -59,6 +60,13 @@ type Sink struct {
 	// httpOut while still serving stdio would corrupt the MCP protocol
 	// channel.
 	httpMode bool
+
+	// mu makes the writer-swap + file-close (in reload and Close) mutually
+	// exclusive with in-flight Write calls, so a descriptor is never closed
+	// mid-write. That race is a harmless EBADF on POSIX but deadlocks the
+	// runtime poller on Windows. Write holds the shared read lock (writers
+	// stay concurrent); reload and Close take the exclusive lock.
+	mu sync.RWMutex
 
 	// writer is what every Write call routes through. Updated by reload via
 	// Store; never mutated in place.
@@ -132,8 +140,11 @@ func New(cfg *config.StaticConfig, httpOut, errOut io.Writer) (*Sink, error) {
 }
 
 // Write implements io.Writer. It routes to whichever writer Reload most
-// recently installed.
+// recently installed, holding mu's read lock so a reload cannot close the
+// file mid-write (see mu).
 func (s *Sink) Write(p []byte) (int, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
 	h := s.writer.Load()
 	if h == nil {
 		return io.Discard.Write(p)
@@ -183,6 +194,9 @@ func (s *Sink) Reload(cfg *config.StaticConfig) error {
 // returns an error) lands somewhere visible instead of being swallowed by
 // the file descriptor we are about to close.
 func (s *Sink) Close() error {
+	// Exclusive lock so the swap+close can't race an in-flight Write (see mu).
+	s.mu.Lock()
+	defer s.mu.Unlock()
 	s.writer.Store(&writerHolder{s.errOut})
 	old := s.file.Swap(nil)
 	if old == nil {
@@ -233,19 +247,24 @@ func (s *Sink) applyDestination(cfg *config.StaticConfig) error {
 
 	s.lastLogFile = cfg.LogFile
 	s.applied = true
-	s.writer.Store(&writerHolder{newWriter})
 
-	// In-flight writes that already loaded the old writer may still hit the
-	// previous fd briefly after Close — same race log rotation already
-	// accepts (logrotate's copytruncate has the identical property).
+	// Swap + close under the exclusive lock (see mu). The new file was opened
+	// above, outside the lock, so writers only ever block for the swap+close,
+	// not for open latency.
+	s.mu.Lock()
+	s.writer.Store(&writerHolder{newWriter})
 	oldFile := s.file.Swap(newFile)
+	var closeErr error
 	if oldFile != nil {
-		if err := oldFile.Close(); err != nil {
-			// Surface this at default verbosity — a Close failure on
-			// rotation is the kind of thing that drops fds on disk-pressure
-			// or flaky network filesystems.
-			klog.Warningf("logging: failed to close previous log file: %v", err)
-		}
+		closeErr = oldFile.Close()
+	}
+	s.mu.Unlock()
+
+	if closeErr != nil {
+		// Logged outside the lock: klog.Warningf re-enters Write's read lock
+		// and sync.RWMutex is not reentrant. A rotation Close failure warrants
+		// default verbosity — it drops fds under disk pressure or flaky netfs.
+		klog.Warningf("logging: failed to close previous log file: %v", closeErr)
 	}
 	return nil
 }

--- a/pkg/logging/sink_test.go
+++ b/pkg/logging/sink_test.go
@@ -6,6 +6,7 @@ import (
 	"path/filepath"
 	"runtime"
 	"sync"
+	"sync/atomic"
 	"testing"
 
 	"github.com/stretchr/testify/suite"
@@ -304,6 +305,54 @@ func (s *SinkSuite) TestConcurrentWriteAndReloadIsRaceFree() {
 	}
 	close(stop)
 	wg.Wait()
+}
+
+func (s *SinkSuite) TestConcurrentWriteAndReloadNeverWritesToClosedFile() {
+	// Reload closes the previously-open file inline (sink.go applyDestination).
+	// Without mutual exclusion between Write and that close, a Write that has
+	// already loaded the old writer can land on the just-closed descriptor: on
+	// POSIX that surfaces as an os.ErrClosed write error, and on Windows it can
+	// wedge the runtime poller and deadlock the whole package (issue #1199).
+	// Drive direct Sink.Write (bypassing klog's global serialization, so the
+	// writers are genuinely concurrent) against a tight Reload loop and assert
+	// that no write ever observes a closed descriptor.
+	pathA := filepath.Join(s.tempDir, "a.log")
+	pathB := filepath.Join(s.tempDir, "b.log")
+	sink := s.newSink(&config.StaticConfig{LogLevel: 1, LogFile: pathA})
+
+	stop := make(chan struct{})
+	var writeErr atomic.Value // first non-nil error observed by any writer
+	var wg sync.WaitGroup
+	for i := 0; i < 8; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for {
+				select {
+				case <-stop:
+					return
+				default:
+					if _, err := sink.Write([]byte("hot-loop\n")); err != nil {
+						writeErr.CompareAndSwap(nil, err)
+					}
+					runtime.Gosched()
+				}
+			}
+		}()
+	}
+
+	for i := 0; i < 2000; i++ {
+		path := pathA
+		if i%2 == 1 {
+			path = pathB
+		}
+		s.Require().NoError(sink.Reload(&config.StaticConfig{LogLevel: 1, LogFile: path}))
+	}
+	close(stop)
+	wg.Wait()
+
+	err, _ := writeErr.Load().(error)
+	s.NoError(err, "a concurrent Write landed on a descriptor that Reload had already closed")
 }
 
 func (s *SinkSuite) TestReloadIgnoresPortChangeForServeMode() {

